File: Remove Nodes From Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNodes(head *ListNode) *ListNode {
    stack := make([]*ListNode, 0)
    ptr := head

    for ptr != nil {
        for len(stack) > 0 && stack[len(stack)-1].Val < ptr.Val {
            stack = stack[:len(stack)-1]
        }
        stack = append(stack, ptr)
        ptr = ptr.Next
    }

    for i := 0; i < len(stack)-1; i++ {
        stack[i].Next = stack[i+1]
    }
    return stack[0]
}

File: Remove Nodes From Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNodes(ListNode* head) {
        vector<ListNode*> stack;
        ListNode* ptr = head;

        while (ptr != nullptr) {
            while (stack.size() > 0 && stack.back()->val < ptr->val) {
                stack.pop_back();
            }

            stack.push_back(move(ptr));
            ptr = ptr->next;
        }

        for (int i = 0; i < stack.size()-1; i++) {
            stack[i]->next = stack[i+1];
        }
        return stack[0];
    }
};
